Remove MDP from PersistedClusterStateService

rjernst opened the pull request:
This commit removes multiple data path support from the
PersistedClusterStateService. Most notably, many error cases are no
longer possible where duplicate/conflicting data could previously exist
across multiple paths, there is now only one path.

relates #71205

rjernst committed:
Remove MDP from PersistedClusterStateService

This commit removes multiple data path support from the
PersistedClusterStateService. Most notably, many error cases are no
longer possible where duplicate/conflicting data could previously exist
across multiple paths, there is now only one path.

relates #71205

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst commented:
Note that to avoid an enormous PR removing MDP support, this PR tries to narrowly remove support from a single service. There are many examples where the calling code creates a service for testing with multiple paths, but just uses the first path. This is ok for as an intermediate step since we already block any actual uses of MDP in real nodes.

rjernst committed:
checkstyle

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java:
```diff
@@ -376,9 +321,6 @@ private OnDiskState loadOnDiskState(Path dataPath, DirectoryReader reader) throw
             final Metadata metadata = Metadata.Builder.fromXContent(XContentFactory.xContent(XContentType.SMILE)
                 .createParser(namedXContentRegistry, LoggingDeprecationHandler.INSTANCE, bytes.bytes, bytes.offset, bytes.length));
             logger.trace("found global metadata with last-accepted term [{}]", metadata.coordinationMetadata().term());
-            if (builderReference.get() != null) {
```
This change seems unrelated, it's here to catch a case where someone meddled with the metadata and left us with two copies of the global metadata in the same data path. I think we should keep this.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java:
```diff
@@ -388,16 +330,11 @@ private OnDiskState loadOnDiskState(Path dataPath, DirectoryReader reader) throw
             final IndexMetadata indexMetadata = IndexMetadata.fromXContent(XContentFactory.xContent(XContentType.SMILE)
                 .createParser(namedXContentRegistry, LoggingDeprecationHandler.INSTANCE, bytes.bytes, bytes.offset, bytes.length));
             logger.trace("found index metadata for {}", indexMetadata.getIndex());
-            if (indexUUIDs.add(indexMetadata.getIndexUUID()) == false) {
```
This change seems unrelated, it's here to catch a case where someone meddled with the metadata and left us with two copies of the metadata for the same index in the same data path. I think we should keep this.

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/PersistedClusterStateServiceTests.java:
```diff
@@ -490,94 +284,13 @@ public void testFailsIfGlobalMetadataIsMissing() throws IOException {
         }
     }
 
-    public void testFailsIfGlobalMetadataIsDuplicated() throws IOException {
```
This test is still meaningful, although we'll need to work out a more creative way to duplicate that document.

DaveCTurner replied:
Here is the same test but using two distinct `NodeEnvironment` objects to get the metadata indices that we mash together. This works on `master` as it stands today, obviously needs adjusting for the single-data-path case.

```
    public void testFailsIfGlobalMetadataIsDuplicated() throws IOException {
        // if someone attempted surgery on the metadata index by hand, e.g. deleting broken segments, then maybe the global metadata
        // is duplicated

        final Path[] dataPaths1 = createDataPaths();
        final Path[] dataPaths2 = createDataPaths();

        try (NodeEnvironment nodeEnvironment1 = newNodeEnvironment(dataPaths1);
             NodeEnvironment nodeEnvironment2 = newNodeEnvironment(dataPaths2)) {
            try (Writer writer1 = newPersistedClusterStateService(nodeEnvironment1).createWriter();
                 Writer writer2 = newPersistedClusterStateService(nodeEnvironment2).createWriter()) {
                final ClusterState oldClusterState = loadPersistedClusterState(newPersistedClusterStateService(nodeEnvironment1));
                final long newVersion = randomLongBetween(1L, Long.MAX_VALUE);
                final ClusterState newClusterState = ClusterState.builder(oldClusterState).version(newVersion).build();
                writeState(writer1, 0L, newClusterState, oldClusterState);
                writeState(writer2, 0L, newClusterState, oldClusterState);
            }

            final Path brokenPath = randomFrom(nodeEnvironment1.nodeDataPaths());
            final Path dupPath = randomFrom(nodeEnvironment2.nodeDataPaths());
            try (Directory directory = new SimpleFSDirectory(brokenPath.resolve(PersistedClusterStateService.METADATA_DIRECTORY_NAME));
                 Directory dupDirectory = new SimpleFSDirectory(dupPath.resolve(PersistedClusterStateService.METADATA_DIRECTORY_NAME))) {
                try (IndexWriter indexWriter = new IndexWriter(directory, new IndexWriterConfig())) {
                    indexWriter.addIndexes(dupDirectory);
                    indexWriter.commit();
                }
            }

            final String message = expectThrows(IllegalStateException.class,
                () -> newPersistedClusterStateService(nodeEnvironment1).loadBestOnDiskState()).getMessage();
            assertThat(message, allOf(containsString("duplicate global metadata found"), containsString(brokenPath.toString())));
        }
    }
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/PersistedClusterStateServiceTests.java:
```diff
@@ -490,94 +284,13 @@ public void testFailsIfGlobalMetadataIsMissing() throws IOException {
-        }
-    }
-
-    public void testFailsIfIndexMetadataIsDuplicated() throws IOException {
```
This test is still meaningful, although we'll need to work out a more creative way to duplicate that document.

DaveCTurner replied:
Here is the same test but using two distinct `NodeEnvironment` objects to get the metadata indices that we mash together. This works on `master` as it stands today, obviously needs adjusting for the single-data-path case.

```
    public void testFailsIfIndexMetadataIsDuplicated() throws IOException {
        // if someone attempted surgery on the metadata index by hand, e.g. deleting broken segments, then maybe some index metadata
        // is duplicated

        final Path[] dataPaths1 = createDataPaths();
        final Path[] dataPaths2 = createDataPaths();

        try (NodeEnvironment nodeEnvironment1 = newNodeEnvironment(dataPaths1);
             NodeEnvironment nodeEnvironment2 = newNodeEnvironment(dataPaths2)) {
            final String indexUUID = UUIDs.randomBase64UUID(random());
            final String indexName = randomAlphaOfLength(10);

            try (Writer writer1 = newPersistedClusterStateService(nodeEnvironment1).createWriter();
                 Writer writer2 = newPersistedClusterStateService(nodeEnvironment2).createWriter()) {
                final ClusterState oldClusterState = loadPersistedClusterState(newPersistedClusterStateService(nodeEnvironment1));
                final ClusterState newClusterState = ClusterState.builder(oldClusterState)
                        .metadata(Metadata.builder(oldClusterState.metadata())
                                .version(1L)
                                .coordinationMetadata(CoordinationMetadata.builder(oldClusterState.coordinationMetadata()).term(1L).build())
                                .put(IndexMetadata.builder(indexName)
                                        .version(1L)
                                        .settings(Settings.builder()
                                                .put(IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), 1)
                                                .put(IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0)
                                                .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT)
                                                .put(IndexMetadata.SETTING_INDEX_UUID, indexUUID))))
                        .incrementVersion().build();
                writeState(writer1, 0L, newClusterState, oldClusterState);
                writeState(writer2, 0L, newClusterState, oldClusterState);
            }

            final Path brokenPath = randomFrom(nodeEnvironment1.nodeDataPaths());
            final Path dupPath = randomFrom(nodeEnvironment2.nodeDataPaths());
            try (Directory directory = new SimpleFSDirectory(brokenPath.resolve(PersistedClusterStateService.METADATA_DIRECTORY_NAME));
                 Directory dupDirectory = new SimpleFSDirectory(dupPath.resolve(PersistedClusterStateService.METADATA_DIRECTORY_NAME))) {
                try (IndexWriter indexWriter = new IndexWriter(directory, new IndexWriterConfig())) {
                    indexWriter.deleteDocuments(new Term("type", "global")); // do not duplicate global metadata
                    indexWriter.addIndexes(dupDirectory);
                    indexWriter.commit();
                }
            }

            final String message = expectThrows(IllegalStateException.class,
                () -> newPersistedClusterStateService(nodeEnvironment1).loadBestOnDiskState()).getMessage();
            assertThat(message, allOf(
                containsString("duplicate metadata found"),
                containsString(brokenPath.toString()),
                containsString(indexName),
                containsString(indexUUID)));
        }
    }
```

DaveCTurner reviewed (commented):
> This is ok for as an intermediate step since we already block any actual uses of MDP in real nodes.

I left a suggestion at https://github.com/elastic/elasticsearch/pull/72282/files#r621070239 for making this clearer from the code, and a few other observations.

rjernst committed:
Merge branch 'master' into mdp10

rjernst commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -148,24 +148,6 @@ public void start(Settings settings, TransportService transportService, ClusterS
             final long currentTerm = 0L;
             final ClusterState clusterState = prepareInitialClusterState(transportService, clusterService,
                     ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.get(settings)).build());
-            if (persistedClusterStateService.getDataPaths().length > 0) {
```
Oops yes, good catch.

rjernst committed:
restore dup tests and handling code

rjernst committed:
Make Environment.dataFiles singular

The path.data setting is now a singular string, but the method
dataFiles() that gives access to the path is still an array. This commit
renames the method and makes the return type a single Path.

relates #71205

rjernst committed:
checkstyle

rjernst committed:
Merge branch 'master' into mdp12

rjernst committed:
Merge branch 'mdp12' into mdp10

rjernst committed:
Merge branch 'master' into mdp10

rjernst committed:
add back tests and fixup

rjernst commented:
Thanks for laying out the modified tests @DaveCTurner. I've added them back as suggested. I think this PR is ready for review again.

rjernst committed:
checkstyle

DaveCTurner commented on server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandTests.java:
```diff
@@ -63,7 +63,7 @@ public void createNodePaths() throws IOException {
             nodePaths = new Path[] { nodeEnvironment.nodeDataPath() };
             final String nodeId = randomAlphaOfLength(10);
-            try (PersistedClusterStateService.Writer writer = new PersistedClusterStateService(nodePaths, nodeId,
+            try (PersistedClusterStateService.Writer writer = new PersistedClusterStateService(nodePaths[0], nodeId,
```
We can fix `nodePaths` here too, indeed it's only used in this one method so can just be a local variable of type `Path`.

DaveCTurner reviewed (approved):
LGTM, I noted scope for one tiny extra cleanup too but no need for another round.

rjernst committed:
simplify
